File: apps/web/lib/insurance-age.js
/**
 * Asia/Seoul 이용일 기준 보험나이. 칩 표시용이며 sessionStorage에는 넣지 않는다.
 * 달력 월 연산만 쓰고 182일 환산은 쓰지 않는다.
 */

function daysInMonth(year, month) {
  return new Date(year, month, 0).getDate();
}

function anniversaryInYear(birth, year) {
  if (birth.month === 2 && birth.day === 29 && daysInMonth(year, 2) < 29) {
    return { year, month: 2, day: 28 };
  }
  return { year, month: birth.month, day: birth.day };
}

function civilLte(a, b) {
  if (a.year !== b.year) {
    return a.year < b.year;
  }
  if (a.month !== b.month) {
    return a.month < b.month;
  }
  return a.day <= b.day;
}

/**
 * YYYY-MM-DD만 받는다. 잘못된 형식은 null.
 */
export function parseCivilDate(iso) {
  if (typeof iso !== "string" || !/^\d{4}-\d{2}-\d{2}$/.test(iso)) {
    return null;
  }
  const year = Number(iso.slice(0, 4));
  const month = Number(iso.slice(5, 7));
  const day = Number(iso.slice(8, 10));
  if (month < 1 || month > 12 || day < 1 || day > daysInMonth(year, month)) {
    return null;
  }
  return { year, month, day };
}

export function formatCivilDate(civil) {
  const month = String(civil.month).padStart(2, "0");
  const day = String(civil.day).padStart(2, "0");
  return `${civil.year}-${month}-${day}`;
}

/**
 * 브라우저 현재 시각을 Asia/Seoul 달력일로 본다.
 */
export function seoulToday(now = new Date()) {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Seoul",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(now);
  const get = (type) => parts.find((part) => part.type === type)?.value;
  return parseCivilDate(`${get("year")}-${get("month")}-${get("day")}`);
}

export function addCalendarMonths(value, months) {
  const monthIndex = value.month - 1 + months;
  const year = value.year + Math.floor(monthIndex / 12);
  const month = ((monthIndex % 12) + 12) % 12 + 1;
  const day = Math.min(value.day, daysInMonth(year, month));
  return { year, month, day };
}

function lastBirthdayOnOrBefore(birth, asOf) {
  const thisYear = anniversaryInYear(birth, asOf.year);
  if (civilLte(thisYear, asOf)) {
    return thisYear;
  }
  return anniversaryInYear(birth, asOf.year - 1);
}

function attainedAge(birth, asOf) {
  let years = asOf.year - birth.year;
  if (!civilLte(anniversaryInYear(birth, asOf.year), asOf)) {
    years -= 1;
  }
  return years;
}

/**
 * 상령일(마지막 생일+6개월 당일)부터 만나이+1.
 */
export function insuranceAge(birth, asOf) {
  if (!civilLte(birth, asOf)) {
    throw new Error("future_birth");
  }
  const years = attainedAge(birth, asOf);
  const sixMonths = addCalendarMonths(lastBirthdayOnOrBefore(birth, asOf), 6);
  if (civilLte(sixMonths, asOf)) {
    return years + 1;
  }
  return years;
}

/**
 * 칩용 보험나이. 미래 생일이면 표시하지 않는다.
 */
export function insuranceAgeFromIso(birthIso, asOf = seoulToday()) {
  const birth = parseCivilDate(birthIso);
  if (!birth || !asOf) {
    return null;
  }
  try {
    return insuranceAge(birth, asOf);
  } catch {
    return null;
  }
}

File: apps/api/app/insurance_age.py
"""이용일(Asia/Seoul) 기준 보험나이와 실손·자동차·생명 어댑터. 생년월일은 호출 측에서 버린다."""

from __future__ import annotations

import calendar
from datetime import date

AUTO_SAMPLE_INSUFFICIENT = "표본부족"


def add_calendar_months(value: date, months: int) -> date:
    """달력 월을 더하고 월말·윤년은 그 달의 마지막 날로 맞춘다."""
    month_index = value.month - 1 + months
    year = value.year + month_index // 12
    month = month_index % 12 + 1
    day = min(value.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)


def _anniversary_in_year(birth: date, year: int) -> date:
    try:
        return birth.replace(year=year)
    except ValueError:
        return date(year, 2, 28)


def last_birthday_on_or_before(birth: date, as_of: date) -> date:
    """as_of 이전(당일 포함)의 가장 최근 생일. 2월 29일은 평년에 28일로 본다."""
    this_year = _anniversary_in_year(birth, as_of.year)
    if this_year <= as_of:
        return this_year
    return _anniversary_in_year(birth, as_of.year - 1)


def attained_age(birth: date, as_of: date) -> int:
    """as_of 기준 만나이. 생일이 아직이면 한 살을 빼다."""
    years = as_of.year - birth.year
    if as_of < _anniversary_in_year(birth, as_of.year):
        years -= 1
    return years


def insurance_age(birth: date, as_of: date) -> int:
    """상령일(마지막 생일+6개월 당일)부터 만나이+1. 182일 환산은 쓰지 않는다."""
    if birth > as_of:
        raise ValueError("future_birth")
    years = attained_age(birth, as_of)
    six_months = add_calendar_months(last_birthday_on_or_before(birth, as_of), 6)
    if as_of >= six_months:
        return years + 1
    return years


def age_band_auto(age: int) -> str:
    """자동차 캐시 aggr 문자열. 0–9세는 표본 부족 안내용 값이다."""
    if age < 10:
        return AUTO_SAMPLE_INSUFFICIENT
    if age < 30:
        return "20대 이하"
    if age < 40:
        return "30대"
    if age < 50:
        return "40대"
    if age < 60:
        return "50대"
    if age < 70:
        return "60대"
    return "70대 이상"


def age_band_life(age: int) -> str:
    """생명 캐시 rchn_aggr 문자열."""
    if age < 10:
        return "10세 미만"
    if age < 20:
        return "10대"
    if age < 30:
        return "20대"
    if age < 40:
        return "30대"
    if age < 50:
        return "40대"
    if age < 60:
        return "50대"
    if age < 70:
        return "60대"
    if age < 80:
        return "70대"
    return "80세 이상"
